src/swap/saga.ts

```diff
@@ -0,0 +1,109 @@
+    const priceDiff: number = yield call(
+      getPriceDiff,
+      +data.payload.unvalidatedSwapTransaction.price,
+      +data.payload.unvalidatedSwapTransaction.guaranteedPrice
+    )
+    if (priceDiff >= 2) {
+      yield put(swapPriceChange())
+      return
+    }
+
+    // Set contract kit, wallet address and normalizer
+    const kit: ContractKit = yield call(getContractKit)
+    const walletAddress: string = yield call(getConnectedUnlockedAccount)
+    const normalizer = new TxParamsNormalizer(kit.connection)
+
+    // Approve transaction
+    yield put(swapApprove())
+    Logger.debug(TAG, `Starting to swap approval for address: ${walletAddress}`)
+    const rawApproveTx = { ...data.payload.approveTransaction, from: walletAddress }
+    applyChainIdWorkaround(rawApproveTx, yield call([kit.connection, 'chainId']))
+    const approveTx: CeloTx = yield call(normalizer.populate.bind(normalizer), rawApproveTx)
+    const approveTxo = buildTxo(kit, approveTx)
+    yield call(
+      sendTransaction,
+      approveTxo,
+      walletAddress,
+      newTransactionContext(TAG, 'Swap/Approve')
+    )
+
+    // Query the execute swap endpoint
+    const amountType: string = data.payload.userInput.buyAmount ? 'buyAmount' : 'sellAmount'
+    const amount = data.payload.unvalidatedSwapTransaction[amountType]
+    const params = {
+      buyToken: data.payload.unvalidatedSwapTransaction.buyTokenAddress,
+      sellToken: data.payload.unvalidatedSwapTransaction.sellTokenAddress,
+      [amountType]: amount,
+      userAddress: walletAddress,
+    }
+    const queryParams = new URLSearchParams({ ...params }).toString()
+    const requestUrl = `${networkConfig.executeSwapUrl}?${queryParams}`
+    const response: Response = yield call(fetchWithTimeout, requestUrl)
+    if (!response.ok) {
+      Logger.error(TAG, `Swap failed with status: ${response.status}`)
+      yield put(swapError())
+      return
+    }
+    const responseJson: { validatedSwapTransaction: any } = yield call([response, 'json'])
+
+    // Execute transaction
+    yield put(swapExecute())
+    Logger.debug(TAG, `Starting to swap execute for address: ${walletAddress}`)
+    const rawExecuteTx = responseJson.validatedSwapTransaction
+    applyChainIdWorkaround(rawExecuteTx, yield call([kit.connection, 'chainId']))
+    const executeTx: CeloTx = yield call(normalizer.populate.bind(normalizer), rawExecuteTx)
+    const executeTxo = buildTxo(kit, executeTx)
+    yield call(
+      sendTransaction,
+      executeTxo,
+      walletAddress,
+      newTransactionContext(TAG, 'Swap/Execute')
+    )
+
+    // TODO Tomm: should we check that transaction was successful on the chain e.g the transaction isn't reverted.
+    yield put(swapSuccess())
+  } catch (error) {
+    Logger.error(TAG, 'Error while swapping', error)
+    yield put(swapError())
+  }
+}
+
+export function* swapSaga() {
+  Logger.debug(TAG, 'Initializing swap sagas')
+  yield takeEvery(swapStart.type, swapSubmitSaga)
```

Comment:
`yield takeLatest(swapStart.type, swapStartSaga)`? :) i think that `takeLatest` would work better here, we don't need to kick off multiple of these

Reply:
Fixed in https://github.com/valora-inc/wallet/pull/2891/commits/5c99dfc5b28195ce5357e72cf7c1c1399d09f383!